README.rst
sqlalchemy_tzaware
==================

A project to create a timezone-aware sqlalchemy composite column, to enable any database
to hold datetime information that sorts correctly (by UTC), and displays
correctly (with the original timezone).

This code was originally created against sqlalchemy_ 0.6, and has *not* been
verified to work against 0.7.

.. _sqlalchemy: http://www.sqlalchemy.org/


HISTORY.txt
========================
tzaware_datetime history
========================

v0.5.0
+++++++

- Refactored helper to require target column names.

v0.4.0
+++++++

Initial release of API.



test_tzaware_datetime.py
#!/usr/bin/env python
# -*- coding: iso-8859-1 -*-
"""tests for sqlalchemy timezone-aware datetime support

Tests:
- test class without db access
- test class within db access
"""
# stdlib
import unittest
import datetime

# 3rd-party
import dateutil

# sqlalchemy
from sqlalchemy import MetaData, Table, Column, DateTime, Unicode, Integer
from sqlalchemy import create_engine
from sqlalchemy.orm import mapper, relation, composite, create_session, clear_mappers
from sqlalchemy.orm import CompositeProperty

# module to test
import tzaware_datetime

class TestBasicClass(unittest.TestCase):
    def test_basic_usage(self):
        """Basic usage"""
        # new empty value
        a = tzaware_datetime.TZAwareDateTime()
        self.assertEqual(a.realdate, None)
        b = tzaware_datetime.TZAwareDateTime()
        self.assertEqual(a.realdate, b.realdate)

    def test_roundtrip_UTC(self):
        """Compare tzutc & .utcdt values"""
        newdate = datetime.datetime(2010, 1, 15, 8, tzinfo=dateutil.tz.tzutc())
        newtzadt = tzaware_datetime.TZAwareDateTime(utcdt = newdate)
        self.assertEqual(newdate, newtzadt.realdate)
        self.assertEqual(newdate, newtzadt.utcdt)
    
    def test_roundtrip_NZST(self):
        """Roundtrip between New Zealand standard time"""
        newdate = datetime.datetime(2010, 1, 15, 8, tzinfo=dateutil.tz.tzstr('NZST'))
        newtzadt = tzaware_datetime.TZAwareDateTime(realdate=newdate)
        self.assertEqual(newdate, newtzadt.realdate)
        self.assertEqual(newdate.astimezone(dateutil.tz.tzutc()), newtzadt.utcdt)
    
    def test_roundtrip_PST(self):
        """Roundtrip between Pacific standard time"""
        newdate = datetime.datetime(2010, 1, 15, 8, tzinfo=dateutil.tz.gettz('PST'))
        newtzadt = tzaware_datetime.TZAwareDateTime(realdate=newdate)
        self.assertEqual(newdate, newtzadt.realdate)
        
    def test_retrieve_all_properties(self):
        """Retrieve all properties when only one is set"""
        newdate = datetime.datetime(2010, 1, 15, 9, 30, tzinfo=dateutil.tz.gettz('EST'))
        newtzadt = tzaware_datetime.TZAwareDateTime()
        self.assertTrue(newtzadt.utcdt == newtzadt.tzname == newtzadt.offsetseconds == newtzadt.realdate == None,
                        'Everything should be None')
        newtzadt.realdate = newdate
        self.assertFalse(newtzadt.utcdt == None)
        self.assertFalse(newtzadt.realdate == None)
        self.assertFalse(newtzadt.offsetseconds == None)
        
        newtzadt_naive = tzaware_datetime.TZAwareDateTime(utcdt=datetime.datetime(2010, 1, 15, 9, 45))
        self.assertFalse(newtzadt_naive.realdate == None, 
                         '.realdate from naive datetime should NOT be None')
        self.assertEqual(newtzadt_naive.realdate, 
                         datetime.datetime(2010, 1, 15, 9, 45, 
                                           tzinfo=dateutil.tz.tzutc()),
                         'Naive datetime should generate UTC timestamp')
    
    def test_compare_dates_from_different_timezones(self):
        """Compare dates from different timezones"""
        newdate_Sydney = datetime.datetime(2010, 1, 15, 8, 
                                           tzinfo=dateutil.tz.gettz('Australia/Sydney'))
        newdate_Rio = datetime.datetime(2010, 1, 15, 8, 
                                        tzinfo=dateutil.tz.gettz('America/Sao_Paulo'))
        newtzadt_Sydney = tzaware_datetime.TZAwareDateTime(realdate=newdate_Sydney)
        newtzadt_Rio = tzaware_datetime.TZAwareDateTime(realdate=newdate_Rio)
        self.assertNotEqual(newtzadt_Sydney.realdate, newtzadt_Rio.realdate)
        self.assertNotEqual(newtzadt_Sydney.utcdt, newtzadt_Rio.utcdt)
        
    def test_compare_offsets(self):
        """Compare offsets from datetime and TZAwareDateTime objects"""
        newdate_a = datetime.datetime(2010, 1, 15, 8, tzinfo=dateutil.tz.gettz('Europe/Rome'))
        newtzadt_a = tzaware_datetime.TZAwareDateTime(realdate=newdate_a)
        
        # Brazil official time (BRT, aka UTC-3)
        newdate_b = datetime.datetime(2010, 1, 15, 8, 
                                      tzinfo=dateutil.tz.gettz('America/Sao Paulo'))
        newtzadt_b = tzaware_datetime.TZAwareDateTime(realdate=newdate_b)

        # compare offsets
        newdate_a_offset = self.calc_offset(newdate_a.utcoffset())
        self.assertEqual(newdate_a_offset, newtzadt_a.offsetseconds)
        
        self.assertEqual(self.calc_offset(newdate_b.utcoffset()), 
                         newtzadt_b.offsetseconds)
        
    def calc_offset(self, tdelta):
        """properly calculates an offset in seconds (negative when day is 0, positive when day is 1)"""
        if tdelta.days == 0:
            return -tdelta.seconds
        else:
            return 86400 - tdelta.seconds

class TestDatabaseAccess(unittest.TestCase):
    """Test TZAwareDateTime within a sqlalchemy database"""
    def setUp(self):
        """prep db access"""
        # create engine
        self.db_myengine = create_engine('sqlite:///:memory:', echo=False)
        
        # setup table metadata
        self.db_metadata = MetaData()
        table_infomatic = Table('infomatic', self.db_metadata,
                          Column('id', Integer, primary_key=True),
                          Column('info', Unicode(255)),
                          Column('expectedoffset', Integer),
                          Column('utcdate', DateTime), # for tzawaredate
                          Column('tzname', Unicode), # for tzawaredate
                          Column('tzoffset', Integer)) # for tzawaredate
        
        # setup mappings
        mapper(InfoMatic, table_infomatic, properties={
            'info': table_infomatic.c.info,
            'expectedoffset': table_infomatic.c.expectedoffset,
            'tzawaredate': composite(tzaware_datetime.TZAwareDateTime, 
                                     table_infomatic.c.utcdate, 
                                     table_infomatic.c.tzname,
                                     table_infomatic.c.tzoffset)
        })
        
        # create all tables
        self.db_metadata.create_all(self.db_myengine)
        
        # create session
        self.session = create_session(bind=self.db_myengine, autocommit=True, autoflush=True)
    
    def tearDown(self):
        """properly clear sqlalchemy in-memory values"""
        self.session.close()
        #self.db_metadata.clear()
        clear_mappers()
        
    def test_roundtrip_date(self):
        """DB: Roundtrip a simple date"""
        newdate = datetime.datetime(2010, 1, 16, 9, tzinfo=dateutil.tz.tzutc())
        newtzadt = tzaware_datetime.TZAwareDateTime(realdate=newdate)
        infomatic = InfoMatic(u'first date', newtzadt)
        self.session.add(infomatic)
        
        newtzadt_fromdb = self.session.query(InfoMatic).first()
        self.assertEqual(newdate, newtzadt_fromdb.tzawaredate.realdate)
        
    def test_sorting(self):
        """add multiple TZAwareDateTimes, return by UTC order"""
        dates_to_add = {'Rome': [u'Rome (UTC+1)', datetime.datetime(2010, 1, 20, 6,
                                                          tzinfo=dateutil.tz.gettz('Europe/Rome'))], 
                        'London': [u'London (UTC)', datetime.datetime(2010, 1, 20, 6,
                                                           tzinfo=dateutil.tz.gettz('Europe/London'))], 
                        'Toronto': [u'Toronto (UTC-5)', datetime.datetime(2010, 1, 20, 6,
                                                              tzinfo=dateutil.tz.gettz('America/Toronto'))]
                        }
        for k, v in dates_to_add.iteritems():
            self.session.add(InfoMatic(v[0], tzaware_datetime.TZAwareDateTime(realdate=v[1])))
            
        dates_in_order = self.session.query(InfoMatic).order_by(InfoMatic.tzawaredate).all()
        
        # check for all dates
        self.assertTrue(len(dates_in_order) == 3)
        # check for order
        self.assertTrue(dates_in_order[0].tzawaredate.realdate == dates_to_add['Rome'][1])
        self.assertTrue(dates_in_order[2].tzawaredate.realdate == dates_to_add['Toronto'][1])
        
class TestDatabaseSetupHelper(unittest.TestCase):
    """test the TZAwareDateTime sqlalchemy setup helpers"""
    def setUp(self):
        pass
    
    def tearDown(self):
        clear_mappers()

    def test_add_columns(self):
        """Add Column objects"""
        # setup table metadata
        db_metadata = MetaData()
        table_infomatic = Table('infomatic', db_metadata,
                          Column('id', Integer, primary_key=True),
                          Column('info', Unicode(255)),
                          Column('expectedoffset', Integer))
        self.assertTrue(len(table_infomatic.columns) == 3)
        
        # append 3 TZADT columns
        tzaware_datetime.helper.append_columns(table_infomatic, 'newdate')
        
        self.assertTrue(len(table_infomatic.columns) == 6)
        
    def test_add_multiple_columns(self):
        """Add multiple tzadt columns"""
        # setup table metadata
        db_metadata = MetaData()
        table_x = Table('x', db_metadata,
                        Column('id', Integer, primary_key=True),
                        Column('info', Unicode(255)),
                        Column('expectedoffset', Integer))
        self.assertTrue(len(table_x.columns) == 3)
        
        # append 3 TZADT columns
        tzaware_datetime.helper.append_columns(table_x, 'onedate')
        self.assertEqual(len(table_x.columns), 6)
        
        # append 3 more TZADT columns
        tzaware_datetime.helper.append_columns(table_x, 'twodate')
        self.assertEqual(len(table_x.columns), 9)
        
    def test_set_mapping(self):
        """set mappings"""
        # setup table metadata
        db_metadata = MetaData()
        datecolumnname = 'newdate'
        table_infomatic = Table('infomatic', db_metadata,
                          Column('id', Integer, primary_key=True),
                          Column('info', Unicode(255)),
                          Column('expectedoffset', Integer))
        
        # append 3 TZADT columns
        tzaware_datetime.helper.append_columns(table_infomatic, datecolumnname)
        
        # get the mapper dictionary
        column_defintion = tzaware_datetime.helper.get_mapper_definition(table_infomatic, datecolumnname)
        
        self.assertTrue(isinstance(column_defintion, CompositeProperty))
 
    def test_set_mapping_multiple(self):
        """set mappings for multiple TZAwareDateTime columns"""
        # setup table metadata
        db_metadata = MetaData()
        datecolumnnames = ('reddate', 'bluedate')
        table_infomatic = Table('infomatic', db_metadata,
                          Column('id', Integer, primary_key=True),
                          Column('info', Unicode(255)),
                          Column('expectedoffset', Integer))
        
        # append 2 sets of TZADT columns
        tzaware_datetime.helper.append_columns(table_infomatic, datecolumnnames[0])
        tzaware_datetime.helper.append_columns(table_infomatic, datecolumnnames[1])
        
        # get the mapper dictionaries
        for datecolumnname in datecolumnnames:
            columndefinition = tzaware_datetime.helper.get_mapper_definition(table_infomatic, datecolumnname)
            self.assertTrue(isinstance(columndefinition, CompositeProperty))

    def test_whole_enchilada(self):
        """test entire database"""
        # create engine
        db_myengine = create_engine('sqlite:///:memory:', echo=False)
        
        # setup table metadata
        db_metadata = MetaData()
        table_infomatic = Table('infomatic', db_metadata,
                          Column('id', Integer, primary_key=True),
                          Column('info', Unicode(255)),
                          Column('expectedoffset', Integer))
        tzadtcolumnname = 'thedate'
        tzaware_datetime.helper.append_columns(table_infomatic, tzadtcolumnname)
        
        # setup mappings
        mapper(InfoMatic, table_infomatic, properties={
            'info': table_infomatic.c.info,
            'expectedoffset': table_infomatic.c.expectedoffset,
            tzadtcolumnname: tzaware_datetime.helper.get_mapper_definition(table_infomatic, tzadtcolumnname)
        })
        
        # create all tables
        db_metadata.create_all(db_myengine)
        
        # create session
        session = create_session(bind=db_myengine, autocommit=True, autoflush=True)
    
        # properly clear sqlalchemy in-memory values
        session.close()
        
class InfoMatic(object):
    """table to hold TZAwareDateTime values"""
    def __init__(self, info, tzawaredate):
        self.info = info
        self.tzawaredate = tzawaredate
    def __repr__(self):
        return "<InfoMatic('%s', %s)" % (self.info, self.tzawaredate)

def run_all_tests():
    # create a complete suite of tests
    allsuites = unittest.TestSuite()

    # add all TestCase subclasses
    allsuites.addTests(unittest.TestLoader().loadTestsFromTestCase(TestBasicClass))
    allsuites.addTests(unittest.TestLoader().loadTestsFromTestCase(TestDatabaseAccess))
    allsuites.addTests(unittest.TestLoader().loadTestsFromTestCase(TestDatabaseSetupHelper))
    unittest.TextTestRunner(verbosity=2).run(allsuites)

if __name__ == '__main__':
    run_all_tests()


tzaware_datetime.py
#!/usr/bin/env python
# -*- coding: iso-8859-1 -*-
"""sqlalchemy timezone-aware datetime support

Usage:
  Add the following columns to the table definition:
    Column('utcdate', DateTime),
    Column('tzname', Unicode),
    Column('tzoffset', Integer))

  In the mapper, add the following key:
  'tzawaredate': composite(TZAwareDateTime, 
                            thetable.c.utcdate, 
                            thetable.c.tzname,
                            thetable.c.tzoffset)
                            
  The columns can be named anything, but they must exist with those types and be reference in that order.
"""
__version_info__ = ('0', '5', '0')
__version__ = '.'.join(__version_info__)
__author__ = 'Andrew Ittner <aji@rhymingpanda.com>'
# stdlib
from datetime import datetime, timedelta

# sqlalchemy
from sqlalchemy import Column, DateTime, Unicode, Integer
from sqlalchemy.orm import composite

# dateutil <http://labix.org/python-dateutil>
from dateutil import tz

# module-level data
TZAwareDateTimeColumns = (Column('utcdate', DateTime),
                          Column('tzname', Unicode),
                          Column('tzoffset', Integer))
TZAwareDateTimeColumnNames = ('utcdate', 'tzname', 'tzoffset')

class TZAwareDateTime(object):
    """A composite sqlalchemy column that round-trips timezone-aware datetime objects"""
    def __init__(self, utcdt=None, tzname=None, offsetseconds=None, realdate=None):
        """utcdt: UTC datetime
        tzname: human-readable timezone name
        offsetseconds: seconds between local date and UTC
        realdate: actual date in target timezone
        """
        if (realdate is None):
            self.utcdt = utcdt
            self.tzname = tzname
            self.offsetseconds = offsetseconds
        else:
            self._set_realdate(realdate)
            
    def __repr__(self):
        if self.tzname:
            return "<TZDateTime (%s, tzname=%s, offset=%s)>" % (self.realdate, 
                                                                self.tzname,
                                                                self.offsetseconds)
        else:
            return "<TZDateTime (%s, offset=%s)>" % (self.realdate, 
                                                     self.offsetseconds)
        
    def __composite_values__(self):
        return [self.utcdt, self.tzname, self.offsetseconds]

    def __set_composite_values__(self, utcdt, tzname, offsetseconds):
        self.utcdt = utcdt
        self.tzname = tzname
        self.offsetseconds = offsetseconds
        
    def __eq__(self, other):
        return other.utcdt == self.utcdt
    
    def __ne__(self, other):
        return not self.__eq__(other)

    def _calc_offset_seconds(self, tdelta):
        """calculate timedelta seconds based on day value"""
        assert isinstance(tdelta, timedelta)
        return (-tdelta.days * 86400) - tdelta.seconds
    
    def _get_realdate(self):
        """reconstruct timezone-aware date from 3 columns"""
        tz_reconstitute = None
        # use offset from UTC (timezone name not guaranteed for roundtrip)
        if self.offsetseconds is None:
            # return date as UTC
            if self.utcdt is None:
                return None
            return self.utcdt.replace(tzinfo=tz.tzutc())
        else:
            tz_reconstitute = tz.tzoffset(name=None, offset=self.offsetseconds)
            
        # get naive date in UTC timezone, convert to non-naive, offset to given timezone
        return self.utcdt.replace(tzinfo=tz.tzutc()).astimezone(tz_reconstitute)
    
    def _set_realdate(self, newdate):
        """use a single datetime with a timezone to set class values"""
        # convert to utc
        self.utcdt = newdate.astimezone(tz.tzutc())

        # get timezone name
        newtzname = newdate.tzname()
        if newtzname is not None:
            self.tzname = unicode(newtzname)
        else:
            self.tzname = None

        # set offset
        self.offsetseconds = self._calc_offset_seconds(newdate.utcoffset())
    
    realdate = property(_get_realdate, _set_realdate)

class helper(object):
    """functions to insert TZAwareDateTime into database objects"""

    @staticmethod
    def append_columns(newtable, columnname):
        """given a sqlalchemy Table, add the TZAwareDatetime Column objects to it
        Modifies newtable in place"""
        # set prefixes for these columns
        for c in TZAwareDateTimeColumns:
            # NOTE: a copy must be made first, then attributes set;
            #   setting name & key args in column.copy() does not work
            newcolumn = c.copy()
            #assert isinstance(newcolumn, Column)
            newcolumn.name = '%s_%s' % (columnname, c.name)
            newcolumn.key = '%s_%s' % (columnname, c.key)
            newtable.append_column(newcolumn)
            
    @staticmethod
    def get_mapper_definition(newtable, columnname):
        """Given a Table object, return the Mapper definition for a TZAwareDateTime column"""
        # cycle through columns, find the utcdate, tzname, tzoffset columns
        column_utcdate, column_tzname, column_tzoffset = None, None, None
        new_column_names = {'utcdate': '%s_%s' % (columnname, TZAwareDateTimeColumnNames[0]),
                            'tzname': '%s_%s' % (columnname, TZAwareDateTimeColumnNames[1]),
                            'tzoffset': '%s_%s' % (columnname, TZAwareDateTimeColumnNames[2])
                            }

        for c in newtable.c:
            assert isinstance(c, Column)
            if c.key == new_column_names['utcdate']:
                column_utcdate = c
            elif c.key == new_column_names['tzname']:
                column_tzname = c
            elif c.key == new_column_names['tzoffset']:
                column_tzoffset = c
            if column_utcdate != column_tzname != column_tzoffset != None:
                break

        return composite(TZAwareDateTime,
                         column_utcdate,
                         column_tzname,
                         column_tzoffset)



demo.py
#!/usr/bin/env python
# -*- coding: iso-8859-1 -*-
"""Demonstration of TZAwareDateTime composite column for sqlalchemy"""
# stdlib
from datetime import datetime

# sqlalchemy
from sqlalchemy import MetaData, Table, Column, DateTime, Unicode, Integer
from sqlalchemy import create_engine
from sqlalchemy.orm import mapper, relation, composite, create_session

# timezone-aware composite column
from tzaware_datetime import TZAwareDateTime

# 3rd-party: dateutil <http://labix.org/python-dateutil>
from dateutil import tz

# demonstration parent table
class InfoMatic(object):
    """sqlalchemy main demonstration table: contains basic info, plus a composite TZAwareDateTime column"""
    def __init__(self, info=None, tzawaredate=None, expectedoffset=None):
        self.info = info
        self.tzawaredate = tzawaredate
        self.expectedoffset = expectedoffset
    def __repr__(self):
        return "<InfoMatic('%s', %s, %s)" % (self.info, self.tzawaredate, self.expectedoffset)

def prep_database():
    global myengine

    # create engine
    myengine = create_engine('sqlite:///:memory:', echo=False)
    
    # setup table metadata
    metadata = MetaData()
    table_infomatic = Table('infomatic', metadata,
                      Column('id', Integer, primary_key=True),
                      Column('info', Unicode(255)),
                      Column('expectedoffset', Integer),
                      Column('utcdate', DateTime), # for TZAwareDateTime
                      Column('tzname', Unicode), # for TZAwareDateTime
                      Column('tzoffset', Integer)) # for TZAwareDateTime
    
    # setup mappings
    mapper(InfoMatic, table_infomatic, properties={
        'info': table_infomatic.c.info,
        'expectedoffset': table_infomatic.c.expectedoffset,
        'tzawaredate': composite(TZAwareDateTime, 
                                 table_infomatic.c.utcdate, 
                                 table_infomatic.c.tzname,
                                 table_infomatic.c.tzoffset)
    })

    # create all tables
    metadata.create_all(myengine)

def run_demo():
    """prep the database, create a session, run some example code"""
    global myengine

    prep_database()
    
    # create session
    session = create_session(bind=myengine, autocommit=True, autoflush=True) #autoflush=True: key!
    
    # create & save info objects
    lots_of_dates = [InfoMatic(u"first date", TZAwareDateTime(realdate=datetime.now(tz.tzutc())), 0)]
    lots_of_dates.append(InfoMatic(u"null date", TZAwareDateTime(), None))
    lots_of_dates.append(InfoMatic(u"PST date", 
                                   TZAwareDateTime(realdate=datetime.now(tz.gettz("PST"))),
                                   28800))
    lots_of_dates.append(InfoMatic(u"New Zealand date", 
                                   TZAwareDateTime(realdate=datetime.now(tz.gettz("Pacific/Auckland"),
                                                                         ))))
    session.add_all(lots_of_dates)
    
    # print all objects
    info_count = session.query(InfoMatic).count()
    print '\tAll infomatic objects (%s)' % info_count
    for infomatic in session.query(InfoMatic):
        assert isinstance(infomatic, InfoMatic)
        if infomatic.tzawaredate is not None:
            assert isinstance(infomatic.tzawaredate, TZAwareDateTime)
        print infomatic
        print '\t', infomatic.info
        print '\ttzawaredate.realdate', infomatic.tzawaredate.realdate
        print '\ttzawaredate.utcdt', infomatic.tzawaredate.utcdt
        
    session.close()

if __name__ == '__main__':
    run_demo()
    